File: apps/api/src/projects/scheduling/work-item-hierarchy.ts
// Ported from reno_pm's work_item_hierarchy.dart — only groupByParent and
// parentItemIds are needed server-side for parent/summary aggregation.
// flattenTree/FlatTreeItem is a UI tree-flattening concern for the task
// table's collapse/expand rendering and stays client-side (Flutter) only.

import { SchedulingWorkItem } from './scheduling-types';

/**
 * Groups `items` by their parentId. The `null` key holds every top-level
 * item. Each group is sorted the same way the caller passed them in (callers
 * are expected to pass items pre-sorted by sortOrder, mirroring the source
 * app's WorkItem.sortOrder-based sort — kept minimal here since only
 * min-start/max-end aggregation needs this, which is sort-order-independent).
 */
export function groupByParent(
  items: SchedulingWorkItem[],
): Map<string | null, SchedulingWorkItem[]> {
  const map = new Map<string | null, SchedulingWorkItem[]>();
  for (const item of items) {
    const key = item.parentId;
    if (!map.has(key)) map.set(key, []);
    map.get(key)!.push(item);
  }
  return map;
}

/**
 * IDs of items that have at least one child — i.e. summary/母項目 rows whose
 * schedule is an aggregate of their descendants rather than their own
 * duration/predecessors. Whether an item is a parent is always derived this
 * way, never stored, so a leaf automatically becomes a summary row the
 * moment something is nested under it.
 */
export function parentItemIds(items: SchedulingWorkItem[]): Set<string> {
  const result = new Set<string>();
  for (const item of items) {
    if (item.parentId != null) result.add(item.parentId);
  }
  return result;
}

File: apps/windows_app/lib/core/models/work_item_hierarchy.dart
import 'work_item.dart';

/// Groups [items] by their [WorkItem.parentId]. The `null` key holds every
/// top-level item. Each group is sorted by [WorkItem.sortOrder].
Map<String?, List<WorkItem>> groupByParent(List<WorkItem> items) {
  final map = <String?, List<WorkItem>>{};
  for (final item in items) {
    map.putIfAbsent(item.parentId, () => []).add(item);
  }
  for (final list in map.values) {
    list.sort((a, b) => a.sortOrder.compareTo(b.sortOrder));
  }
  return map;
}

/// IDs of items that have at least one child — i.e. summary/母項目 rows
/// whose schedule is an aggregate of their descendants rather than their
/// own duration/predecessors. Whether an item is a parent is always
/// derived this way, never stored, so a leaf automatically becomes a
/// summary row the moment something is nested under it.
Set<String> parentItemIds(List<WorkItem> items) {
  return {
    for (final item in items)
      if (item.parentId != null) item.parentId!,
  };
}

/// One row of a flattened, depth-first tree traversal — what a tree-style
/// task table or a row-aligned Gantt canvas actually iterates over.
class FlatTreeItem {
  final WorkItem item;
  final int depth;
  final bool hasChildren;

  const FlatTreeItem({
    required this.item,
    required this.depth,
    required this.hasChildren,
  });
}

/// Depth-first flattening of the parent/child tree: every top-level item
/// followed immediately by its descendants (recursively, unlimited depth),
/// each sorted by [WorkItem.sortOrder] within its sibling group. A node in
/// [collapsedIds] contributes its own row but its descendants are skipped,
/// so both the task table and the Gantt canvas can stay in row-for-row sync
/// by flattening with the same collapse state.
List<FlatTreeItem> flattenTree(
  List<WorkItem> items, {
  Set<String> collapsedIds = const {},
}) {
  final grouped = groupByParent(items);
  final parentIds = parentItemIds(items);
  final result = <FlatTreeItem>[];

  void visit(String? parentId, int depth) {
    for (final child in grouped[parentId] ?? const <WorkItem>[]) {
      final hasChildren = parentIds.contains(child.id);
      result.add(
        FlatTreeItem(item: child, depth: depth, hasChildren: hasChildren),
      );
      if (hasChildren && !collapsedIds.contains(child.id)) {
        visit(child.id, depth + 1);
      }
    }
  }

  visit(null, 0);
  return result;
}
